### PR TITLE
Option --dump-coreimp to output imperative-core to JSON

### PR DESCRIPTION
**Pull request by metaleap**
New `purs compile` option `--dump-coreimp`, exactly like the existing `--dump-corefn`, to dump the `CoreImp` AST to a `coreimp.json` output file.

Requested in #876 and also in #711 [in the comments](https://github.com/purescript/purescript/issues/711#issuecomment-74744519)

(The results can be seen in the various `coreimp.json` files throughout [this directory](https://github.com/metaleap/gonad-demos/tree/master/01_hello_stranger/output) --- [one example](https://github.com/metaleap/gonad-demos/blob/master/01_hello_stranger/output/Ohai.Greet/coreimp.json) )

**New modules:**

- `src/Language/PureScript/CoreImp/ToJSON.hs` --- modeled after the existing `src/Language/PureScript/CoreFn/ToJSON.hs`. (Has duplicated the trivial tiny helper functions `identToJSON` and `moduleNameToJSON` as I didn't want to impress changes on to what the existing original exports.)

**Existing modules:**

Merely duplications of the existing `--dump-corefn` option handling in:
- `app/Command/Compile.hs`
- `src/Language/PureScript/Make.hs`
- `src/Language/PureScript/Options.hs`
(Too small & trivial to generalize into helper funcs for now IMHO --- until a 3rd `--dump-xyz` request comes along in the future at least.)

### CODE REVIEW

File: src/Language/PureScript/CoreImp/ToJSON.hs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+subAstToJSONMaybe :: Maybe AST -> Value
+subAstToJSONMaybe (Just ast) =
+    astToJSON ast
+subAstToJSONMaybe Nothing =
+    Null
+
+subASTSingleToJSON :: String -> (Maybe SourceSpan) -> AST -> Value
+subASTSingleToJSON name maybeSrcSpan ast =
+    object [T.pack "sourceSpan" .= toJSON maybeSrcSpan,
```
Comment (kRITZCREEK) [inline src/Language/PureScript/CoreImp/ToJSON.hs]: Does this hit the same issue as the CoreFn change, where it stores the file path in every node?

Comment (metaleap) [reply src/Language/PureScript/CoreImp/ToJSON.hs]: Well with `--source-maps` enabled. Is it an issue to begin with? We're emitting *sizable* to-be-machine-processed json files already.. I can certainly ditch the file path if you guys prefer, personally am hoping I won't need the sourcemappings for my ps2go transpiler anyway

File: src/Language/PureScript/CoreImp/ToJSON.hs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            T.pack name         .= astToJSON ast
+            ]
+
+subASTsListToJSON :: String -> (Maybe SourceSpan) -> [AST] -> Value
+subASTsListToJSON name maybeSrcSpan asts =
+    object [T.pack "sourceSpan" .= toJSON maybeSrcSpan,
+            T.pack name         .= map astToJSON asts
+            ]
+
+subASTsLeftRightToJSON :: String -> (Maybe SourceSpan) -> AST -> AST -> Value
```
[the remark this answers is not in the document]
Comment (metaleap) [reply src/Language/PureScript/CoreImp/ToJSON.hs]: All 3 occurrences removed with last added commit

File: src/Language/PureScript/Make.hs [hunk 3]
```diff
@@ -331,14 +332,17 @@ buildMakeActions outputDir filePathMap foreigns usePrefix =
...
   getOutputTimestamp mn = do
     dumpCoreFn <- asks optionsDumpCoreFn
+    dumpCoreImp <- asks optionsDumpCoreImp
     let filePath = T.unpack (runModuleName mn)
         jsFile = outputDir </> filePath </> "index.js"
         externsFile = outputDir </> filePath </> "externs.json"
         coreFnFile = outputDir </> filePath </> "corefn.json"
-        min3 js exts coreFn
+        coreImpFile = outputDir </> filePath </> "coreimp.json"
+        min3 js exts coreFn coreImp
```
[the remark this answers is not in the document]
Comment (metaleap) [reply src/Language/PureScript/Make.hs]: Done in last added commit

File: src/Language/PureScript/CoreImp/ToJSON.hs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import Language.PureScript.CoreFn.ToJSON as CoreFnToJSON
+import Language.PureScript.CoreImp.AST
+import Language.PureScript.PSString
+
+
+
+
+moduleToJSON :: (Show a) => Version -> Module a -> [AST] -> Value
+moduleToJSON v m a = object [ T.pack "body"    .= map (astToJSON) a
+                            , T.pack "corefn"  .= CoreFnToJSON.moduleToJSON v m -- might be good to "keep a snapshot of the original FP logic of the module" depending on coreimp-processing use-case (others can ignore it)
```
Comment (paf31) [inline src/Language/PureScript/CoreImp/ToJSON.hs]: I don't know how I feel about including the functional core here. What's the use case?

File: src/Language/PureScript/CoreImp/ToJSON.hs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                            , T.pack "modraw"  .= toJSON (show m) -- only way to capture type/ctors info without rewriting corefn, good enough for those needing it badly enough, others can ignore it
```
Comment (paf31) [inline src/Language/PureScript/CoreImp/ToJSON.hs]: What does `modraw` mean? Also, let's not use `show` here.

In general, I'd like to keep this minimal. It's harder to remove features later.

Comment (metaleap) [reply src/Language/PureScript/CoreImp/ToJSON.hs]: Yeah I'll do away with this shortly, accidentally pushed this last commit. Still a bit more in flux than I expected I'm afraid!

File: src/Language/PureScript/CoreImp/ToJSON.hs (hunk 4 above)
Comment (metaleap) [reply src/Language/PureScript/CoreImp/ToJSON.hs]: Ultimately I want to push (later) for an argument "include-everything-you-know-incl-especially-type/ctor-infos" to `CoreFn.ToJSON.moduleToJSON`. I get that the existing `--dump-corefn` output is meant to be small and semi-readable for (typically) troubleshooting use-cases, but for the `--dump-coreimp` I see the use-cases as tools that need the most information possible about the original source. More in-depth type info (ADTs / records) in the current (JS-focused) CoreImp will be very hard to reconstruct from these "untyped" (other than stringlit/boollit/numlit/obj-lit) --- including a more detailed corefn (than the original/current `--dump-corefn` scenario envisions) right there in the coreimp dump will probably help. I guess for now that's too vague to let into the main repo so I'll take it out for now in the next commit, then keep tinkering locally until it all solidifies a bit more.

**Comment by metaleap** [comment]
That last commit 5e9d1fc.. wasn't intended, what a mess, sorry

**Pull request closed without merging**